## 7) 테스트
- 어떤 소프트웨어를 공개하기 전에 제대로 테스트가 필요
- 테스트 코드는 실행단위별로 작성해야 하므로, 2가지 레벨에서 구현할 수 있다.
  - 통합 테스트 : 뷰 하나를 호출하는 것에 대한 테스트 / 즉, 하나의 API, 하나의 HTTP 요청이 우리 서버로 들어오고 나가는데 까지를 거쳐가는 모든 코드들을 다 테스트 / 하나의 기능을 동작시키는 모든 코드들을 테스트 하는 것 
  - 단위 테스트 : 함수나 메소드 하나를 호출하는 것에 대한 테스트


### 단위 테스트 실습
- 먼저 settings.py에 가서 DATABASES 부분에 추가해주자.

```python
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.mysql',
        'HOST': 'tablebookings.c1uzjwt0ilcb.ap-northeast-2.rds.amazonaws.com',
        'PORT': 3306,
        'NAME': 'tablebooking',
        'USER': 'tablebookings',
        'PASSWORD': 'qwaszxqwer1!',
        'TEST': {
            'NAME': 'tablebooking_test'
        }
    }
}
```

- 이렇게 TEST라는 키와 딕셔너리 값을 넣어줘야 한다.

- **그리고 이제 templatetags - booking_tags.py에 있는 convert_status_korean 함수가 제대로 동작하는지를 확인해보자.** (단위 테스트)
  - web 앱 디렉터리 내부에 보면 tests.py라는 파일이 만들어져 있다. 여기에 코드를 입력해서 세팅을 해보자.

```python
from django.test import TestCase
from django.test import Client
from django.urls import reverse

from .models import Booking, Restaurant, Category
from .templatetags.booking_tags import convert_status_korean


class BookingTagsTest(TestCase):
    def setUp(self) -> None:
        self.cases = [
            (Booking.PayStatus.READY, "결제 대기"),
            (Booking.PayStatus.CANCELED, "예약 취소"),
            (Booking.PayStatus.FAILED, "예약 실패"),
            (Booking.PayStatus.PAID, "예약 완료"),
        ]
    
    def test_convert_status_korean(self):
        for case in self.cases:
            result = convert_status_korean(case[0])
            self.assertEqual()
            self.assertEqual(case[1], result)
```

- **우리가 단위 테스트를 하려고 하는 함수인 convert_status_korean를 import로 가지고 온다.** 
- **그리고 booking_tags에 있는 함수들을 호출하는 클래스를 만들어주는데, 이 때 테스트를 하기 위해서는 꼭 TestCase를 import 해야 한다.** 
  - 그리고 그 안에는 setUp이라는 함수를 구현해야 한다. 이 함수는 해당 테스트를 수행하기 직전에 미리 사전에 준비해야되는 사전 작업들을 여기에 정의해둔다. 
  - 그래서 여기에 각각의 케이스에 대해서 매핑되는 규칙들을 정의해주자. 
  - 우리가 convert_status_korean 함수가 작동하는 것에 대해서 의도한 내용들이 있을텐데, 그 의도한 내용을 self.cases에 (Booking.PayStatus.READY, "결제 대기") 이렇게 괄호안에 왼쪽은 input / 오른쪽은 Output으로 해서 튜플로 하나를 정의해서 cases에 넣어준 것이다.
  - 그래서 4가지 결제 상태에 대한 우리의 의도한 내용들을 전부 튜플로 정의해준다.

- setUp이라는 함수는 해당 클래스가 처음에 뜰 때 미리 위와같이 케이스를 정의해주거나 미리 준비해야할 데이터베이스를 정리해두거나 그럴 때 사용하는 메소드이다. 
- 그래서 이 테스트를 django에서 실행시키면 setUp이 무조건 먼저 실행이 되고 -> 그 다음에 내가 추가로 실행할 함수들을 그 밑에다가 정의할 수 있다. 
  - **그래서 보통 python 파일 1개 당 --> test.py의 클래스 1개로 매칭시켜서 테스트를 많이 작성한다. 그래서 해당 파일 1개의 함수의 개수만큼 test.py 클래스 내부에 메소드가 늘어나게 된다.**
  - **그래서 이 클래스가 실행이 될 때, setUp 메소드가 먼저 실행되고 그 밑에 있는, 우리가 정의한 메소드들이 실행되는 것이다.**
  - 밑에 우리가 정의하는 메소드들은 보통 test라는 이름으로 시작하는 것이 좋다.

- **test_convert_status_korean라는 함수 내부에 for cases in self.cases: 이렇게 함으로써 -> setUp에서 미리 정의한 테스트 케이스들을 여기서 쓸 수 있게 되는 것이다.** 
  - 그리고 result = convert_status_korean(case[0]) 이렇게 해서 함수를 호출하고 튜플의 0번을 대입해준다. 
  - self.assertEqual() 이렇게 assertEqual이라는 메소드가 있는데, TestCase에서 기본적으로 제공하고 있는 메소드이다. 이건 첫번째와 두번째가 값이 동일한지를 평가해준다. 
  - **self.assertEqual(case[1], result) 이렇게 내가 예상한 결과(테스트 결과)와 / 실제 함수를 호출했을 때 나오는 값이 동일한지 확인해주자.**

- **그래서 동일하면 문제없이 통과가 되고 / 동일하지 않으면 이 assertEqual에서 테스트가 실패했다는 exception을 발생시킨다.**
  - AssertionException, AssertionError가 발생하고 테스트에 실패하게 된다. 

- 이러한 상태에서 이제 터미널에서 

```python
python manage.py test
```

- 이렇게 입력하게 되면, 

```terminal
Found 1 test(s).
Creating test database for alias 'default'...
System check identified some issues:

----------------------------------------------------------------------
Ran 1 test in 0.053s

OK
Destroying test database for alias 'default'...
```

- **이렇게 테스트에 통과했다고 뜨게 된다.** 
- Creating test database for alias 'default'... 이러한 얘기가 나온다. 우리가 아까 settings.py에서 DATABASES에 TEST로 넣어주었다. 이건, 데이터베이스와 연계된 것들이 많은데, 데이터베이스에 의존적인 코드들이 많을 수 밖에 없다. 그러면 데이터베이스 의존적인 작업들은 어떻게 테스트를 해야되는지 문제가 있다. 
  - 예를 들어서 우리가 테스트가 뜰 때마다 실제 DB에 붙어서 테스팅이 된다고 하면, 실제 서비스에도 영향이 갈 수 있다. 그래서 우리가 더미 데이터를 넣는 그런 작업을 이 테스트 코드에서 setUp에 수행을 해놨는데 이걸 뒤처리까지 다 해줘야 한다. 이렇게 대충 입력한 데이터들이 사용자한테 보여지는 건 문제가 있기 때문이다. 
  - **그래서 보통은 샌드박스라는 환경과 프로덕션 환경을 구분해서 서버를 다르게 배포하기도 하고 / 테스트할 때도 내가 테스트한 정보들이 실제 서비스에 오염되지 않도록 실제로 테스트 방법론에서도, 테스트 DB를 구별해서 작업하는 걸 권장하고 있다.**
  - **그래서 우리가 아까 settings.py - DATABASES에 써준 내용은 테스트 DB로, 우리가 python manage.py test라고 입력해서 서버를 구동하고 DBeaver를 보면 테스트 DB가 생성된 걸 확인할 수 있다. 그리고 터미널에서 테스트가 끝났다고 나오는 순간 다시 확인해보면 -> 이 DB가 자동으로 삭제가 되어있다.**
  - 그래서 이런식으로 테스트시에만 사용할 수 있는 DB를 알아서 생성해줬다가 알아서 파괴시켜주는 그런 역할을 하는 게 settings.py - DATABASES 코드이다.

* * *

- **일부러 오류를 내보자.**
- booking_tags.py의 convert_status_korean 함수에서 예약 취소 부분은 "예약 취소!!!" 이렇게 해놓고 python manage.py test를 입력해보면,

```terminal
Found 1 test(s).
Creating test database for alias 'default'...
System check identified some issues:

System check identified 1 issue (0 silenced).
F
======================================================================
FAIL: test_convert_status_korean (web.tests.BookingTagsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kimsangbaek/Desktop/mytasklist/table_bookings/web/tests.py", line 22, in test_convert_status_korean
    self.assertEqual(case[1], result)
AssertionError: '예약 취소' != '예약 취소!!!'
- 예약 취소
+ 예약 취소!!!
?      +++


----------------------------------------------------------------------
Ran 1 test in 0.051s

FAILED (failures=1)
Destroying test database for alias 'default'...
```

- 이런식으로 FAILED가 뜬다. 그리고 뭐가 틀렸는지도 확인해준다. 이렇게 뭐가 잘못되었는지 표시해준다. 


### 통합 테스트 실습
- test.py에 이어서 클래스를 작성해보도록 하자.

```python
from django.test import Client
...

# 통합 테스트 실습
class SearchViewTest(TestCase):
    def setUp(self) -> None:
        self.client = Client()
        self.test_restaurants = []
        category = Category.objects.create(name='한식')
        self.test_restaurants.append(
            Restaurant.objects.create(name='테스트', category=category, address='서울시 중구', phone='020000')
        )
        self.test_restaurants.append(
            Restaurant.objects.create(name='아무거나', category=category, address='서울시 중구', phone='020000')
        )
        
     def test_get(self):
        keyword = '테스트'
        response = self.client.get(reverse('search'), {'keyword': keyword})

        self.assertTemplateUsed(response, 'main/search.html')
        for item in self.test_restaurants:
            if keyword in item.name:
                self.assertContains(response, '<div class="card" id="restaurant-%d">' % item.id, status_code=200)
            else:
                self.assertNotContains(response, '<div class="card" id="restaurant-%d">' % item.id, status_code=200)

```

- 이번에는 실제로 HTTP 요청을 통해서 통합 테스트를 수행해야 하기 때문에 from django.test import Client 이렇게 Client를 가지고 와준다. 
  - 이 Client는 HTTP 요청을 테스트를 위해서 발생시켜주는 역할을 한다. 
  - 그래서 setUp 메소드에 self.client = Client() 이렇게 Client 할당을 먼저 해준다. 

- **그리고 category = Category.objects.create() 이렇게 직접 데이터를 만들어줘야 한다. test DB가 따로 생성되기 때문이다.**
  - 그리고나서 self.test_restaurants 리스트에 -> Restaurant 모델 데이터를 2개 생성해서 카테고리는 바로 위에 걸 설정해서 만들어주기.


- **이제는 setUp 메소드가 아닌 우리가 직접 정의해주는 함수를 만들자.**
  - 우리가 setUp에서 데이터를 넣었을 때, 이 데이터에 맞게 검색결과가 의도한대로 동작하는지를 보는 테스트 코드가 된다.
  - 그래서 def test_get(self) 이렇게 설정하고 -> 이 searchView에서 get이 발생했을 때 의도대로 동작하는지를 보는 테스트가 된다. 
  - response = self.client.get(reverse('search')) 는 우리의 client를 사용해서 search url path를 가지고 와서 키워드 값으로 keyword를 전달해준다. 

- **그래서 SearchViewTest 함수는 이 client가 호출이 될 때, main.py에 있는 SearchView에 keyword = self.request.GET.get('keyword') 이 부분에 테스트라는 값이 들어가게 된다. 이걸 지금 시뮬레이션 한 것이다.** 
- 그럴 경우에, self.assertTemplateUsed(response, 'main/search.html') 해당 코드로 search.html을 제대로 사용하고 있는지를 평가해서 이게 맞으면, self.test_restaurants 리스트에서 1개씩 빼서 해당 데이터 name안에 keyword가 있는지 확인.
  - **그리고 assertContains를 사용해서 restaurant 아이디가 있는 카드인지를 찾아보면 된다.** 

* * *

- 이제 실제로 main 디렉터리 - search.html를 보면, 검색 결과가 card로 뜨는 것을 확인할 수 있다. 근데 우리가 하고 싶은 건, 응답을 html로 받고 카드가 뜰 때 "테스트"라는 키워드가 들어간 상점들만 떠야할 것이다. 그래서 카드에서 id값으로, 그 id가 검색결과에 존재하면 맞는 것이고 존재하면 안되는 건데 하드에 뜨는 경우는 실패했다는 것을 표시해주고 싶다. 

- 그래서 div class="card" 부분에 id값 코드를 추가해준다.

```html
...
<h4>검색 결과</h4>
...
<div class="card" id="restaurant-{{ restaurant.id }}">

```

- 이렇게 테스트를 위해서 이 카드가 누구것인지 식별하기 위해 pk를 부여해주는 것이다. 


- **다시 test.py로 가서 코드를 보면, response를 검증할 때, keyword에 속해있는 레스토랑이라면 이 pk가 있어야 한다 / 그리고 만약에 속하지 않아야 되는 레스토랑이라면 assertNotContains 조건을 만족해야 된다는 것이다.**
- 이제 다시 python manage.py test로 실행해보면 된다. 그러면 둘 다 잘 되었다고 나온다.
